presto/dbapi.py

```diff
@@ -236,10 +245,111 @@ def setinputsizes(self, sizes):
+        # one that is going to be used to execute the actual operation.
+        query = presto.client.PrestoQuery(copy.deepcopy(self._request), sql=sql)
+        result = query.execute()
+
+        # Iterate until the 'X-Presto-Added-Prepare' header is found or
+        # until there are no more results
+        for _ in result:
+            response_headers = result.response_headers
+
+            if constants.HEADER_ADDED_PREPARE in response_headers:
+                return response_headers[constants.HEADER_ADDED_PREPARE]
+
+        raise presto.exceptions.FailedToObtainAddedPrepareHeader
+
+    def _get_added_prepare_statement_presto_query(
+        self,
+        added_prepare_header,
+        statement_name,
+        params
+    ):
+        sql = 'EXECUTE ' + statement_name + ' USING ' + ','.join(map(str, params))
+
+        # No need to deepcopy _request here because this is the actual request
+        # operation
+        return presto.client.PrestoQuery(self._request, sql=sql)
+
+    def _execute_deallocate_prepare(self, added_prepare_header, statement_name):
+        sql = 'DEALLOCATE PREPARE ' + statement_name
+
+        # Send deallocate statement. Copy the _request object to avoid poluting the
+        # one that is going to be used to execute the actual operation.
+        query = presto.client.PrestoQuery(copy.deepcopy(self._request), sql=sql)
+        result = query.execute(
+            additional_http_headers={
+                constants.HEADER_PREPARED_STATEMENT: added_prepare_header
+            }
+        )
+
+        # Iterate until the 'X-Presto-Deallocated-Prepare' header is found or
+        # until there are no more results
+        for _ in result:
+            response_headers = result.response_headers
+
+            if constants.HEADER_DEALLOCATED_PREPARE in response_headers:
+                return response_headers[constants.HEADER_DEALLOCATED_PREPARE]
+
+        raise presto.exceptions.FailedToObtainDeallocatedPrepareHeader
+
+    def execute(self, operation, params=None):
+        if params:
+            assert isinstance(params, (list, tuple)), (
+                'params must be a list or tuple containing the query '
+                'parameter values'
+            )
+
+            # Generate a unique statement name as convinient way to avoid
+            # passing it as a parameter to the current function
+            statement_name = 'st_' + uuid.uuid4().hex.replace('-', '')
+
+            # Send prepare statement
+            added_prepare_header = self._execute_prepare_statement(
+                operation, statement_name
+            )
+
+            # Send execute statement and assign the return value to `results`
+            # is it will be returned by the function
+            self._query = self._get_added_prepare_statement_presto_query(
+                added_prepare_header, statement_name, params
+            )
+            result = self._query.execute(
+                additional_http_headers={
```

Comment:
I was thinking that maybe we should change headers map in  `self._query`. But passing it explicitly here seems fine too.